com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
+                    res.append(',').append(Signature.getTypeVariable(typeParameters[i]));
+                }
+                res.append('>');
+            }
+        }
+        return res.toString();
     }
 
-    private void doHotCodeReplace(List<IResource> resources, List<String> fullyQualifiedName) {
-        // DODO: Add HCR logic
-        synchronized (needHotCodeReplace) {
-            needHotCodeReplace.set(false);
-            needHotCodeReplace.notifyAll();
+    private String getTypeName(String genericTypeSignature) {
+        int arrayDimension = 0;
+        while (genericTypeSignature.charAt(arrayDimension) == '[') {
+            arrayDimension++;
+        }
+        int parameterStart = genericTypeSignature.indexOf('<');
+        StringBuffer name = new StringBuffer();
+        if (parameterStart < 0) {
+            name.append(genericTypeSignature.substring(arrayDimension + 1, genericTypeSignature.length() - 1)
+                    .replace('/', '.'));
+        } else {
+            if (parameterStart != 0) {
+                name.append(genericTypeSignature.substring(arrayDimension + 1, parameterStart).replace('/', '.'));
+            }
+            try {
+                String sig = Signature.toString(genericTypeSignature)
+                        .substring(Math.max(parameterStart - 1, 0) - arrayDimension);
+                name.append(sig.replace('/', '.'));
+            } catch (IllegalArgumentException iae) {
+                // do nothing
+                name.append(genericTypeSignature);
+            }
         }
+        for (int i = 0; i < arrayDimension; i++) {
+            name.append("[]"); //$NON-NLS-1$
+        }
+        return name.toString();
+    }
+
+    private boolean supportsDropToFrame(ThreadReference thread, StackFrame frame) {
+        List<StackFrame> frames = getStackFrames(thread, false);
+        for (int i = 0; i < frames.size(); i++) {
+            if (StackFrameUtility.isNative(frames.get(i))) {
+                return false;
+            }
+            if (frames.get(i) == frame) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected void popStackFrame(StackFrame frame) throws DebugException {
+        if (frame != null) {
+            ThreadReference thread = frame.thread();
+            List<StackFrame> frames = getStackFrames(thread, false);
+            int desiredSize = frames.indexOf(frame);
+            while (desiredSize >= 0) {
+                StackFrameUtility.pop(frames.get(0));
+                frames = getStackFrames(thread, true);
+                desiredSize--;
+            }
+        }
+    }
+
+    protected void dropToStackFrame(StackFrame frame) throws DebugException {
+        // Pop the drop frame and all frames above it
+        popStackFrame(frame);
+        stepIntoThread(frame.thread());
+    }
+
+    private void redefineTypesJDK(List<IResource> resources, List<String> qualifiedNames) throws DebugException {
```

Comment:
Why is the function name ends with JDK?

Reply:
There is other redefinTypes other than the JDK version, such as IBM's J9. Just use this naming convention from JDT's implementation.